File: core/src/main/java/dev/vibris/core/TerminalResult.java
package dev.vibris.core;

import dev.vibris.protocol.v1.JobCompleted;
import dev.vibris.protocol.v1.JobFailed;
import dev.vibris.protocol.v1.ProtocolVersion;
import dev.vibris.protocol.v1.ServerMessage;

record TerminalResult(JobCompleted completed, JobFailed failed) {
    private static final ProtocolVersion V1 = ProtocolVersion.newBuilder().setMajor(1).setMinor(0).build();

    static TerminalResult completed(JobCompleted value) {
        return new TerminalResult(value, null);
    }

    static TerminalResult failed(JobFailed value) {
        return new TerminalResult(null, value);
    }

    boolean successful() {
        return completed != null;
    }

    ServerMessage message(String messageId, String requestId, String workspaceId) {
        ServerMessage.Builder result = ServerMessage.newBuilder()
            .setProtocolVersion(V1)
            .setMessageId(messageId)
            .setRequestId(requestId)
            .setWorkspaceId(workspaceId);
        return successful() ? result.setJobCompleted(completed).build() : result.setJobFailed(failed).build();
    }
}

File: core/src/main/kotlin/dev/vibris/core/TerminalResult.kt
package dev.vibris.core

import dev.vibris.protocol.v2.JobCompleted
import dev.vibris.protocol.v2.JobFailed
import dev.vibris.protocol.v2.ProtocolVersion
import dev.vibris.protocol.v2.ServerMessage

@JvmRecord
internal data class TerminalResult(
    val completed: JobCompleted?,
    val failed: JobFailed?,
) {
    fun successful(): Boolean = completed != null

    fun message(messageId: String, requestId: String, workspaceId: String): ServerMessage {
        val result = ServerMessage.newBuilder()
            .setProtocolVersion(V2)
            .setMessageId(messageId)
            .setRequestId(requestId)
            .setWorkspaceId(workspaceId)
        return if (successful()) {
            result.setJobCompleted(completed).build()
        } else {
            result.setJobFailed(failed).build()
        }
    }

    companion object {
        private val V2 = ProtocolVersion.newBuilder().setMajor(2).setMinor(0).build()

        @JvmStatic
        fun completed(value: JobCompleted): TerminalResult = TerminalResult(value, null)

        @JvmStatic
        fun failed(value: JobFailed): TerminalResult = TerminalResult(null, value)
    }
}
